### PR TITLE
BaiChuan2 Model

### PR DESCRIPTION
**Pull request by blacker521**
Support for the BaiChuan2 Generative Model

## Motivation
Baichuan2 Model Support

## Checklist

- [x] Format your code according to the [Contributor Guide](https://github.com/sgl-project/sglang/blob/main/docs/en/contributor_guide.md).
- [x] Add unit tests as outlined in the [Contributor Guide](https://github.com/sgl-project/sglang/blob/main/docs/en/contributor_guide.md).
- [x] Update documentation as needed, including docstrings or example tutorials.

### CODE REVIEW

File: python/sglang/srt/models/baichuan.py [hunk 1]
```diff
@@ -0,0 +1,442 @@
...
+    ):
+        if config.hidden_size == 4096:  # baichuan2 7b
+            super().__init__(config, "ROPE", cache_config, quant_config,
+                             lora_config)
+        else:  # baichuan 13b, baichuan2 13b
+            super().__init__(config, "ALIBI", cache_config, quant_config,
+                             lora_config)
+
+
+class BaiChuanForCausalLM(BaiChuanBaseForCausalLM):
```
Comment (merrymercy) [inline python/sglang/srt/models/baichuan.py]: remove the duplicated classes

**Comment by merrymercy** [comment]
please fix the lint https://github.com/sgl-project/sglang/blob/main/docs/en/contributor_guide.md

File: python/sglang/srt/models/baichuan.py (hunk 1 above)
Comment (blacker521) [reply python/sglang/srt/models/baichuan.py]: Is not the same class

File: python/sglang/srt/models/baichuan.py [hunk 2]
```diff
@@ -0,0 +1,448 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Inference-only BaiChuan model compatible with HuggingFace weights."""
+import math
+from typing import Any, Iterable, List, Optional, Tuple
+
+import torch
+from torch import nn
+from transformers import PretrainedConfig
+from vllm.config import CacheConfig, LoRAConfig
```
Comment (zhyncs) [inline python/sglang/srt/models/baichuan.py]: remove LoRA related imports

File: python/sglang/srt/models/baichuan.py [hunk 3]
```diff
@@ -0,0 +1,448 @@
...
+    RowParallelLinear,
+)
+from vllm.model_executor.layers.quantization.base_config import QuantizationConfig
+from vllm.model_executor.layers.rotary_embedding import get_rope
+from vllm.model_executor.layers.vocab_parallel_embedding import (
+    ParallelLMHead,
+    VocabParallelEmbedding,
+)
+from vllm.model_executor.model_loader.weight_utils import default_weight_loader
+from vllm.model_executor.models.interfaces import SupportsLoRA
```
Comment (zhyncs) [inline python/sglang/srt/models/baichuan.py]: remove LoRA related imports

File: python/sglang/srt/models/baichuan.py [hunk 4]
```diff
@@ -0,0 +1,448 @@
...
+from vllm.model_executor.sampling_metadata import SamplingMetadata
```
Comment (zhyncs) [inline python/sglang/srt/models/baichuan.py]: We don't need the SamplingMetadata in SGLang

File: python/sglang/srt/models/baichuan.py [hunk 5]
```diff
@@ -0,0 +1,448 @@
...
+
+import torch
+from torch import nn
+from transformers import PretrainedConfig
+from vllm.config import CacheConfig, LoRAConfig
+from vllm.distributed import (
+    get_tensor_model_parallel_rank,
+    get_tensor_model_parallel_world_size,
+)
+from vllm.model_executor.layers.activation import SiluAndMul
```
Comment (zhyncs) [inline python/sglang/srt/models/baichuan.py]: Why do you import SiluAndMul from vLLM?

**Review by zhyncs — CHANGES_REQUESTED** [review]

**Comment by zhyncs** [comment]
Test correctness by comparing the final logits and outputs of the two following commands:
```
python3 scripts/playground/reference_hf.py --model [new model]
python3 -m sglang.bench_latency --model [new model] --correct --output-len 16 --trust-remote-code
```

File: python/sglang/srt/models/baichuan.py (hunk 1 above)
Comment (merrymercy) [reply python/sglang/srt/models/baichuan.py]: their class names are the same. Am I missing something?

**Comment by blacker521** [comment]
## python3 scripts/playground/reference_hf.py --model Baichuan2-7B-Chat
========== Prompt 0 ==========
prefill logits (final) tensor([13.9219, 13.6406, 26.0000,  ..., 13.8047, 13.7969, 14.1328],
       device='cuda:0', dtype=torch.float16)
 The capital of France is Paris.
Paris is the capital city of France.</s>

========== Prompt 1 ==========
prefill logits (final) tensor([14.7891, 14.4219, 28.0000,  ..., 14.5781, 14.2344, 14.8281],
       device='cuda:0', dtype=torch.float16)
The capital of the United Kindom is London.
The currency used in the United Kingdom is the pound sterling, symbol

========== Prompt 2 ==========
prefill logits (final) tensor([18.1875, 18.3125, 27.0625,  ..., 18.2031, 18.2188, 18.1250],
       device='cuda:0', dtype=torch.float16)
Today is a sunny day and I like to go for a walk in the park. The flowers are blooming, and
## python3 -m sglang.bench_latency --model Baichuan2-7B-Chat --correct --output-len 16 --trust-remote-code
input_ids=[[1524, 6067, 1376, 7212, 1414], [1524, 6067, 1376, 1352, 2988, 22934, 1384, 1414], [11672, 1414, 1346, 38767, 2332, 1377, 1406, 1795]]

prefill logits (first half): tensor([[24.0000, 23.8750, 39.2500,  ..., 24.0000, 23.8750, 23.7500],
        [ 7.9375,  7.8125, 15.1250,  ...,  8.0000,  8.0625,  8.0625],
        [15.2500, 15.3125, 25.7500,  ..., 15.4375, 15.1875, 15.4375]],
       device='cuda:0') 

prefill logits (final): tensor([[13.8125, 13.5625, 25.8750,  ..., 13.7500, 13.7500, 14.0625],
        [14.7500, 14.3750, 27.8750,  ..., 14.5625, 14.1875, 14.8125],
        [18.1250, 18.3750, 26.8750,  ..., 18.2500, 18.2500, 18.1250]],
       device='cuda:0') 

========== Prompt 0 ==========
The capital of France is Paris.
The capital of France is Paris. The country is located in Western Europe 

========== Prompt 1 ==========
The capital of the United Kindom is London.
The currency of the United Kindom is the pound sterling (GBP 

========== Prompt 2 ==========
Today is a sunny day and I like to go for a walk in the park. The flowers are blooming and the birds

**Comment by blacker521** [comment]
Should "chat_template.py"  be updated?

File: python/sglang/srt/models/baichuan.py (hunk 1 above)
Comment (blacker521) [reply python/sglang/srt/models/baichuan.py]: same as https://github.com/vllm-project/vllm/blob/main/vllm/model_executor/models/baichua
for Baichuan 7B. or Baichuan 13B and Baichuan2 7B/13B.

**Comment by zhyncs** [comment]
@blacker521 Please merge the latest main and resolve the conflicts.

File: python/sglang/srt/models/baichuan.py (hunk 1 above)
Comment (merrymercy) [reply python/sglang/srt/models/baichuan.py]: You have two classes. Their names are the same. The later one will override the first one. Is this correct?

**Review by merrymercy — CHANGES_REQUESTED** [review]

**Comment by merrymercy** [comment]
@blacker521 if you want to use sglang frontend, you should also update `chat_template.py`

**Comment by blacker521** [comment]
I solved the issues,the chat_template.py will be updated over time

**Review by merrymercy — APPROVED** [review]
